binary_test.go

```diff
@@ -37,11 +35,119 @@ func TestBinary(t *testing.T) {
-		t.Fatalf("Binary content from Find is inncorrect. Should be `Hello world!`, was '%s'", str)
+		t.Fatalf("Binary content from FindBinary is incorrect. Should be `Hello world!`, was '%s'", str)
+	}
+}
+
+func TestBinaryKDBXv31RemoveBinary(t *testing.T) {
+	db := NewDatabase()
+
+	db.AddBinary([]byte("test 1"))
+	binary2 := db.AddBinary([]byte("test 2"))
+	binary3 := db.AddBinary([]byte("test 3"))
+	db.AddBinary([]byte("test 4"))
+	db.AddBinary([]byte("test 5"))
+
+	if len(db.Content.Meta.Binaries) != 5 {
+		t.Fatalf("Expected 5 binary elements, found %d", len(db.Content.Meta.Binaries))
+	}
+
+	found := db.FindBinary(binary2.ID)
+	if data, _ := found.GetContentBytes(); string(data) != "test 2" {
+		t.Fatalf("Binary content from FindBinary is incorrect. Should be `test 2`, was '%s'", string(data))
+	}
+
+	removed := db.RemoveBinary(binary2.ID)
+	str, _ := removed.GetContentString()
+	expectedStr, _ := binary2.GetContentString()
+	if str != expectedStr {
+		t.Fatalf(
+			"Binary content from RemoveBinary is incorrect. Should be `%s`, was '%s'",
+			expectedStr,
+			str,
+		)
+	}
+
+	if db.FindBinary(binary2.ID) != nil {
+		t.Fatalf("Binary content from FindBinary is incorrect. It should be removed, but it still exists")
+	}
+
+	if db.FindBinary(binary3.ID) == nil {
+		t.Fatalf("Binary content from FindBinary is incorrect. It should exist")
+	}
+
+	if len(db.Content.Meta.Binaries) != 4 {
+		t.Fatalf("Expected 4 binary elements, found %d", len(db.Content.Meta.Binaries))
+	}
+}
+
+func TestBinaryKDBXv4(t *testing.T) {
+	db := NewDatabase(WithDatabaseKDBXVersion4())
+
+	randomData := make([]byte, 1024*1024)
+	rand.Read(randomData)
+	binary := db.AddBinary(randomData)
+
+	db.LockProtectedEntries()
+	var buffer bytes.Buffer
+	encoder := NewEncoder(&buffer)
+	encoder.Encode(db)
+	db = NewDatabase(WithDatabaseKDBXVersion4())
+	decoder := NewDecoder(bytes.NewReader(buffer.Bytes()))
+	decoder.Decode(db)
+	db.UnlockProtectedEntries()
+
+	found := db.Content.InnerHeader.Binaries.Find(binary.ID)
+	if data, _ := found.GetContentBytes(); string(data) != string(randomData) {
+		t.Log("Received:", len(data))
+		t.Log("Expexted:", len(randomData))
+		t.Fatalf("Binary content from Find is incorrect")
+	}
+}
+
+func TestBinaryKDBXv4RemoveBinary(t *testing.T) {
+	db := NewDatabase(WithDatabaseKDBXVersion4())
+
+	db.AddBinary([]byte("test 1"))
+	binary2 := db.AddBinary([]byte("test 2"))
+	binary3 := db.AddBinary([]byte("test 3"))
+	db.AddBinary([]byte("test 4"))
+	db.AddBinary([]byte("test 5"))
+
+	if len(db.Content.InnerHeader.Binaries) != 5 {
+		t.Fatalf("Expected 5 binary elements, found %d", len(db.Content.InnerHeader.Binaries))
+	}
+
+	found := db.FindBinary(binary2.ID)
+	if data, _ := found.GetContentBytes(); string(data) != "test 2" {
+		t.Fatalf("Binary content from FindBinary is incorrect. Should be `test 2`, was '%s'", string(data))
+	}
+
+	removed := db.RemoveBinary(binary2.ID)
+	str, _ := removed.GetContentString()
```

Comment:
The reason for that appears to be that in KDBX4 the ID is actually not encoded. So when you decode it, it starts to count again at ID 0. If you are still referencing a different ID, it causes the issue.

Not sure how intentional that part is (I actually did not add that part of the KDBXv4 format myself)

Do you happen to have an example KDBXv4 file at hand that was created with Keepass2? (I do not have a windows machine at hand for the next few days)

Reply:
Yes, please. Password enabled, but empty
[Database.zip](https://github.com/tobischo/gokeepasslib/files/8772868/Database.zip).

Reply:
I created empty database, it should be kdbx4.

Reply:
Ah, understood now, my bad. Please use this one
[Database.zip](https://github.com/tobischo/gokeepasslib/files/8773370/Database.zip).

Reply:
Yes, I do need attachments inside to check if the attachment encoding and decoding is working as it does with regular keepass

Reply:
Awesome, thanks

From the data I got from the header there does not actually seem to be an explicit mapping from an ID to the data

Digging a bit into the original source (and I do not claim to understand C#), my understanding so far is, that on loading, the attachments are added to the entry objects.

That means adding and removing attachments is kind of abstracted away to the entry. When writing the keepass file, the binary references are swapped in with the corresponding number from the header, which makes it ok not to have a specific number with the entry, because the reverse is done on load.

If you'd add the same file 5 times, then of course it could still resolve that and reduce the file size 

That would probably be a bigger change to work with that sort of abstraction around it.

The somewhat simpler option I would see is to update the binary reference IDs and binary IDs on binary removal.
That would still not make the test pass, but the encoding and decoding would be correct

Reply:
Thank you for your analysis! I'll try to investigate later today, maybe will find some simple solution.

Reply:
I'm not sure if it is enough to update the IDs on removal because I don't know if it covers all the cases. Alternatively we can set a reference to every Binary struct while reading, hold it inside BinaryReference struct and update IDs while writing. But we need to consider all cases, for example removal some of binaries and then addition few to avoid incorrect IDs before writing.

Also I does'n understood how multiple instances of the same file handled here. You said the approach can help to reduce database size, but it is not clear how library will detect duplication of content (except of gzip compression). Actually I was planning to add possibility to attach binary to a entry with selection of already existing binaries (to avoid duplicates). But if it can be managed automatically, it would be great for user.

Reply:
Alright, so if the binaries are directly attached to the entry when editing the kdbx file, you would have to add or remove the file from there. You wouldn't really care about the ID value, because all you'd need would be a list of binaries.

When encoding the database in order write it back into the file, you'd loop through all entries, add the binary (based on the file version) into the list of binaries we now have. While doing so, you would assign an incrementing ID and create the BinaryReference so that the connection is there.

If multiple entries have the same binary data, then you would just check if the content of any of the files is already identical and if so, take its ID for the binary reference, without storing the file again.

When decoding the file the reverse would be applied:

For each binary reference found with an entry, the binary could be resolved, attached to the entry and thus made editable.

Of course that behaviour would need confirmation with KeePass2.
Based on my looking over the code, that is roughly what is happening there though.

Reply:
Thank you! Now it is much more clear.

Reply:
A different approach could be to just not allow removal of binaries

If you want to delete a binary, you'd just remove the binary reference from the entry.

It would then be up to the encoding step to do the 'garbage collection' (or to not do it and just leave the binary orphaned)

Considering that keepass normally versions its entries it would kind of make sense that instead of removing it, you would rather add the entry to the history and create a new one. That would require the versioned entry to still be able to reference the attachments it has.

Reply:
I am actually thinking, in the context of this pull request, I do not want to add the `RemoveBinary` method, but rather add it at a different time, when this behaviour is better defined.

I will remove it from the commits and then merge it. That at least already solves your issue with the encoding and decoding

Reply:
I like idea with garbage collection, because I planed to implement it by myself in my app. It would be great if the library will provide ready solution.

Also we can extend it to optionally detect and remove duplicates.